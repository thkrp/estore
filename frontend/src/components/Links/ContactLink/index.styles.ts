import styled from 'styled-components';
import { theme } from '../../../styles/theme';

export const ContactLinkStyled = styled.a`
    margin: 5px 10px;
    display: flex;
    align-items: center;
    font-family: ${theme.fonts.semiBold};
    text-decoration: none;
    color: ${theme.colors.grey};
    font-size: 16px;
    line-height: 1;
    cursor: pointer;
    fill: ${theme.colors.grey};
    &:hover {
        color: ${theme.colors.darkgrey};
        fill: ${theme.colors.darkgrey};
    }
`;

export const ContactIconStyled = styled.div`
    display: block;
    width: 15px;
    height: 15px;
    margin-right: 5px;

    & svg {
        width: 100%;
        height: 100%;
    }
`;
