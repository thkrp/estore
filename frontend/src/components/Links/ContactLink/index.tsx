import React, { FC } from 'react';
import { sanitizeUrl } from '@braintree/sanitize-url';
import { ReactComponent as EmailIcon } from '../../../assets/icons/mail.svg';
import { ReactComponent as PhoneIcon } from '../../../assets/icons/old-phone.svg';
import { ContactIconStyled, ContactLinkStyled } from './index.styles';

type Props = {
    value?: string;
    className?: string;
    hiddenIcon?: boolean;
    type: 'phone' | 'email';
};

const Icons = {
    phone: PhoneIcon,
    email: EmailIcon
};

const ContactLink: FC<Props> = ({ value = '', className, hiddenIcon = false, type }) => {
    const format = {
        phone: (tel: string) => `tel:${sanitizeUrl(tel.replace(/[^+\d]/g, ''))}`,
        email: (email: string) => `mailto:${sanitizeUrl(email)}`
    };

    const IconComponent = Icons[type];
    const formatContact = format[type];

    return (
        <ContactLinkStyled href={formatContact(value)} className={className}>
            {!hiddenIcon && (
                <ContactIconStyled>
                    <IconComponent />
                </ContactIconStyled>
            )}
            {value}
        </ContactLinkStyled>
    );
};

export default ContactLink;
