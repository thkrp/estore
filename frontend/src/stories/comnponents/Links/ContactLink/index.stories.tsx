import { Meta, StoryObj } from '@storybook/react';
import { withRouter } from 'storybook-addon-react-router-v6';
// eslint-disable-next-line no-restricted-imports
import ContactLinkComponent from '../../../../components/Links/ContactLink';

const meta = {
    title: 'Components/Links/ContactLink',
    component: ContactLinkComponent,
    decorators: [withRouter],
    parameters: {
        layout: 'centered'
    },
    tags: ['autodocs'],
    argTypes: {
        value: {
            description: 'phone number or email address'
        },
        className: {
            description: 'class names'
        },
        hiddenIcon: {
            description: 'show or hide icon'
        },
        type: {
            description: 'type of link'
        }
    }
} satisfies Meta<typeof ContactLinkComponent>;

export default meta;
type Story = StoryObj<typeof meta>;
export const Email: Story = {
    args: {
        value: 'test@test.com',
        type: 'email'
    }
};

export const Phone: Story = {
    args: {
        value: '+380 50 000 00 00',
        type: 'phone'
    }
};

export const WithoutIcon: Story = {
    args: {
        value: '+380 50 000 00 00',
        type: 'phone',
        hiddenIcon: true
    }
};
