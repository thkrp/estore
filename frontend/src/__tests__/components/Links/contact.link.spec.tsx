/* eslint-disable no-restricted-imports*/
import React from 'react';
import '@testing-library/jest-dom';
import { cleanup, render, screen } from '@testing-library/react';
import ContactLink from '../../../components/Links/ContactLink';

describe('ContactLink', () => {
    afterEach(cleanup);

    it('should be able to pass a class', () => {
        render(<ContactLink value="test@test.com" className="email-link" type="email" />);
        const element = screen.getByRole('link');
        const classes = element.className.split(' ');
        expect(classes.includes('email-link')).toBe(true);
    });

    it('should contain an icon', () => {
        const { container } = render(<ContactLink value="test@test.com" className="email-link" type="email" />);
        expect(screen.getByText(/.*\.(gif|jpe?g|bmp|png|svg)$/i)).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it('should be able to hide an icon', () => {
        const { container } = render(
            <ContactLink value="test@test.com" className="email-link" type="email" hiddenIcon />
        );
        expect(screen.queryByText(/.*\.(gif|jpe?g|bmp|png|svg)$/gi)).not.toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it('should contain a link text', () => {
        render(<ContactLink value="test@test.com" className="email-link" type="email" />);
        expect(screen.getByText(/test@test.com/i)).toBeInTheDocument();
    });

    it('should remove spaces in phone number link', () => {
        render(<ContactLink value="+38 050 000 00 00" type="phone" />);
        const element = screen.getByRole('link');
        expect(element).toHaveAttribute('href', 'tel:+380500000000');
    });
});
