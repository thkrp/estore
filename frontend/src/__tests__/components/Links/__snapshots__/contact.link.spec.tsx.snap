// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactLink should be able to hide an icon 1`] = `
<div>
  <a
    class="sc-bcPKhP bBmFaB email-link"
    href="mailto:test@test.com"
  >
    test@test.com
  </a>
</div>
`;

exports[`ContactLink should contain an icon 1`] = `
<div>
  <a
    class="sc-bcPKhP bBmFaB email-link"
    href="mailto:test@test.com"
  >
    <div
      class="sc-grXZZQ gdnAgs"
    >
      <svg>
        mail.svg
      </svg>
    </div>
    test@test.com
  </a>
</div>
`;
